src/Common/FileCache.h

```diff
@@ -223,8 +224,17 @@ class LRUFileCache final : public IFileCache
     using FileSegmentsByOffset = std::map<size_t, FileSegmentCell>;
     using CachedFiles = std::unordered_map<Key, FileSegmentsByOffset>;
 
+    using AccessKeyAndOffset = std::pair<Key, size_t>;
+    using AccessRecord = std::map<AccessKeyAndOffset, LRUQueue::Iterator>;
```

Comment:
Look like order is not important, so use `std::unordered_map`?

Reply:
> Look like order is not important, so use `std::unordered_map`?

Yes, I tried to use unordered_map before, but std::pair does not support being used as the key of unordered_map (unless use boost::hash , such as unordered_map<pair<string, string>, int, boost::hash<pair<string , string>>>). Can you suggest some modifications?

Reply:
May be you can make it a struct with key and offset fields. 
I did the same here https://github.com/ClickHouse/ClickHouse/blob/35d2dec8d54d5f847f94996efd2c3eb110d61cdb/src/Common/IFileCache.h#L29-L31
In order to make it work with `std::unordered_map` I added this https://github.com/ClickHouse/ClickHouse/blob/35d2dec8d54d5f847f94996efd2c3eb110d61cdb/src/Common/IFileCache.h#L142-L147
But this way probably is not better that your solution with `boost::hash`, why not use it?